hogben/tests/test_optimise.py

```diff
@@ -0,0 +1,134 @@
+    mock_optimise.return_value = np.array([0.8847156, 0.88834418,
+                                           0.00139696,
+                                           0.99860304]), -0.7573710562837207
+    angles, splits, _ = optimiser.optimise_angle_times(num_angles,
+                                                       angle_bounds=(0.2, 2.3),
+                                                       verbose=False)
+    assert len(angles) == num_angles and len(splits) == num_angles
+
+
+@patch('hogben.optimise.Optimiser._Optimiser__optimise')
+def test_optimise_contrasts(mock_optimise):
+    """
+    Tests that the optimise_contrasts method outputs the correct amount of
+    contrasts and counting times.
+    """
+    optimiser = Optimiser(BilayerDMPC())
+    num_contrasts = 3
+    angle_times = [(0.7, 100, 10), (2.3, 100, 40)]
+
+    # Mock values retreived from previous run
+    mock_optimise.return_value = (
+        np.array([-0.56, 2.15, 6.36, 0.17, 0.28, 0.56]), -0.18
+    )
+    contrasts, splits, _ = optimiser.optimise_contrasts(num_contrasts,
+                                                        angle_times,
+                                                        workers=-1,
+                                                        verbose=False)
+    assert len(contrasts) == num_contrasts and len(splits) == num_contrasts
+
+
+@patch('hogben.optimise.Optimiser._Optimiser__optimise')
+def test_optimise_underlayers(mock_optimise):
+    """
+    Tests that the optimise_contrasts method outputs the correct amount of
+    contrasts and counting times.
+    """
+    optimiser = Optimiser(BilayerDMPC())
+
+    num_underlayers = 3
+    angle_times = [(0.7, 100, 10), (2.3, 100, 40)]
+    contrasts = [-0.56, 6.36]
+    thick_bounds = (0, 500)
+    sld_bounds = (1, 9)
+
+    # Mock values retreived from previous run
+    mock_optimise.return_value = (
+        np.array([-0.56, 2.15, 6.36, 0.17, 0.28, 0.56]), -0.18
+    )
+
+    contrasts, splits, _ = optimiser.optimise_underlayers(num_underlayers,
+                                                          angle_times,
+                                                          contrasts,
+                                                          thick_bounds,
+                                                          sld_bounds,
+                                                          verbose=False)
+    assert len(contrasts) == num_underlayers and len(splits) == num_underlayers
+
+
+def test_angle_times_func_result(refnx_sample):
+    """Checks that the angle_times_func method gives the correct result"""
+    x = [0.3, 1.3, 0.8, 0.2]  # [angle, angle, time, time]
+    num_angles = 2
+    contrasts = [3, 14, -2]
+    points = 100
+    total_time = 10000
+
+    optimiser = Optimiser(refnx_sample)
+    result = optimiser._angle_times_func(x, num_angles, contrasts, points,
+                                         total_time)
+    expected_result = -1.7721879778537162
```

Comment:
expected results have quite a large number of digits given you only want a `rtol` of `1e-6` but it really doesn't affect anything so leave it

Reply:
Yeah, I increased the tolerance when I noticed that the test started failing when running in some different order, the issue there lies with the randomness due to the poisson "noise" in the data. So I upped the tolerance and the timing until I got a consistently passing test. The default value for `rtol` is `1e-7` by the way. Will do some testing to see if I can get this to a bit more reasonable values. 

Either way, I agree that so many significant digits make little sense when the tolerance is not in the area of this. So will also amend that a bit.